File: 39. Combination Sum.java
class Solution {
    HashSet<List<Integer>> hs = new HashSet<>();
    public void backtrack(int[] candidates, int index, int target, List<Integer> curr, int currSum){
        if(currSum == target){
            hs.add(new ArrayList<>(curr));
            return;
        }
        if(currSum > target || index >= candidates.length) return;

        curr.add(candidates[index]);
        backtrack(candidates, index, target, curr, currSum + candidates[index]);
        curr.remove(curr.size() - 1);
        
        backtrack(candidates, index + 1, target, curr, currSum);
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        backtrack(candidates, 0, target, new ArrayList<Integer>(), 0);

        List<List<Integer>> ans = new ArrayList<>();
        for(List<Integer> arr : hs) ans.add(arr);

        return ans;
    }
}

File: 39. Combination Sum.cpp
class Solution {
public:
    void backtrack(vector<int>& candidates, int target, vector<vector<int>>& arr,
                   int index, int n, int currSum, vector<int>& currArr) {
        if (index == n) {
            if (currSum == target) {
                arr.push_back(currArr);
            }
            return;
        }

        if (currSum == target) {
            arr.push_back(currArr);
            return;
        }
 
        if (currSum + candidates[index] > target) {
            return;
        }

        currArr.push_back(candidates[index]);
        backtrack(candidates, target, arr, index, n, currSum + candidates[index], currArr);
        currArr.pop_back();

        backtrack(candidates, target, arr, index + 1, n, currSum, currArr);
    }

    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end()); 
        vector<vector<int>> arr;
        vector<int> currArr;
        backtrack(candidates, target, arr, 0, candidates.size(), 0, currArr);
        return arr;
    }
};
